### ISSUE TITLE
Guide people to use pip when bare setup.py used

### ISSUE THREAD

**Issue by maphew** [leo-editor/leo-editor#1547]
suggested by @tfer in thread https://groups.google.com/d/topic/leo-editor/0KLxQ4eEXKw/discussion

Using setup.py by itself to install Leo will not install dependencies, leading to several rounds of trying to launch Leo, note failures, install the failed modules, try launch again. How about adding a message that it's much easier with pip, and perhaps even a guard against running setup.py naively?

**Comment by maphew** [comment]
Yes for sure to the message. I'm not sure about the guard, as I wouldn't want to block downstream packagers who have enough troubles with our current setup.py as is (c.f. @brobr #1512 and others)

**Comment by brobr** [comment]
> Using setup.py by itself to install Leo will not install dependencies, leading to several rounds of trying to launch Leo

This is normal, for any python program. A message about this would be superfluous.
Pip resolves dependencies (beforehand), setup.py won't (be expected to; it just tries to put the stuff into the correct place).

The way around this, is either using pip, as one could do if only managing one computer. When, one has to maintain a couple of computers running (or want to distribute install-scripts for) a particular linux distribution, one would like to use the (format of) the distribution-specific install/upgrade scripts.

Anyone going this way, i.e. providing such a script or packages, should inform users (or the linux-distribution pkg-management tools) what dependencies need to be met.

One way to find this out is as described above; but it is faster to follow what pip does, that is inspecting the `/leo-<version>/leo.egg-info/requires.txt` in the source tarball downloaded from pypi or github. This list dependencies, which can have dependencies on their own, resolved by pip automatically, but that also could already be (made) available for the linux distribution. If the latter, one just installs those packages; if not, one has to create separate packages/buildscripts for each of these dependencies (a bit of work when one does not keep up with the development... ;-)

**Comment by edreamleo** [comment]
@brobr Thanks for your helpful comments.

**Comment by maphew** [comment]
I've decided the current situation is pretty good and there's nothing further to be done in setup.py. Thank you for your comments @brobr.

However it might be good if Leo itself said something. I tried this in launchLeo.py but it looks like the import guard needs to be somewhere else to work:

~~~
#!/usr/bin/env python

""" Leo launcher script
A minimal script to launch leo.
"""
try:
    import leo.core.runLeo
    leo.core.runLeo.run()
except ImportError as e:
    print("\n*** Leo could not be started. Please review the pre-requisite libraries & modulesmodules\n")
    print(e)
~~~

Result with PyQt not available:
~~~
$ python launchLeo.py
leoQt.py: can not fully import PyQt5.
Traceback (most recent call last):

  File "D:\Matt\code\leo-editor\leo\core\leoQt.py", line 52, in <module>
    from PyQt5 import QtCore

ImportError: cannot import name 'QtCore'


Traceback (most recent call last):
  File "launchLeo.py", line 7, in <module>
    import leo.core.runLeo
  File "D:\Matt\code\leo-editor\leo\core\runLeo.py", line 25, in <module>
    g.app = leoApp.LeoApp()
  File "D:\Matt\code\leo-editor\leo\core\leoApp.py", line 338, in __init__
    import leo.core.leoFrame as leoFrame
  File "D:\Matt\code\leo-editor\leo\core\leoFrame.py", line 13, in <module>
    import leo.core.leoColorizer as leoColorizer
  File "D:\Matt\code\leo-editor\leo\core\leoColorizer.py", line 18, in <module>
    from leo.core.leoQt import Qsci, QtGui, QtWidgets
  File "D:\Matt\code\leo-editor\leo\core\leoQt.py", line 82, in <module>
    qt_version = QtCore.QT_VERSION_STR
NameError: name 'QtCore' is not defined
~~~

**Comment by maphew** [comment]
Figured it out, need to catch outer Exception and not just ImportError when the import is nested. Implemented in 00848c4. I put the guard and handler in `runLeo.py` instead of `launchLeo.py` to support `python -m leo.core.runLeo` also.

This is what happens now when something like pyqt isn't available:
~~~
$ python -m leo.core.runLeo
leoQt.py: can not fully import PyQt5.
Traceback (most recent call last):

  File "D:\Matt\code\leo-editor\leo\core\leoQt.py", line 52, in <module>
    from PyQt5 import QtCore

ImportError: cannot import name 'QtCore'


name 'QtCore' is not defined

*** Leo could not be started ***
Please verify you've installed the required dependencies:
https://leoeditor.com/installing.html
~~~

Does this look good to you @edreamleo?

**Comment by brobr** [comment]
Hi Maphew,
I always run Leo from a menu (in Xfce) with a terminal open. This can be set in a 'launcher' or in a `leo.desktop` file; this is the one put in my leo package:
>[Desktop Entry]
Name=Leonine Editor with Outlines (leo)
Name[he]= Leo
Comment=Developer's Text Editor
Exec=leo
Icon=leo
Terminal=false
Type=Application
MimeType=application/x-leo-outline;
Categories=Development;TextEditor

When you set the `Terminal` to `true`, starting leo from a drop-down menu (or desktop-icon) would do this with a terminal open, catching all errors (if any).  This would be the 'ux'-way anyway, something doesn't start from the DE?, start a terminal and run 'something' from the command line. (Maybe this is more for the Windows' side of things? )

Other ways around informing/preparing users (without relying solely on website info, which breaks easily if not automatically updated): make sure that, say, the main README or INSTALL  files are updated with what ends up in the '-egg.info/requires.txt'. Maybe also good to redirect the user to that file (as these come with the package you can make the 'requires' bit to update automatically) in the error-response.

BTW,

This `leo.xml` file is also packaged with the program so that leo documents are linked to leo:
>\<?xml version="1.0"?>
\<mime-info xmlns="http://www.freedesktop.org/standards/shared-mime-info">
\<mime-type type="application/x-leo-outline">
\<comment>leo-outline file</comment>
\<glob pattern="*.leo"/>
\</mime-type>
\</mime-info>

To get the `leo.desktop` and `leo.xml` files immediately used, these commands are run after installing leo:
>if [ -x /usr/bin/update-desktop-database ]; then
   /usr/bin/update-desktop-database -q usr/share/applications >/dev/null 2>&1
fi
>
>if [ -x /usr/bin/update-mime-database ]; then
  /usr/bin/update-mime-database usr/share/mime >/dev/null 2>&1
fi

If you think this could be useful, feel free to include them with leo (although it's specific for linux)

Cheers,
Rob

PS My 'unoffical' leo package for Slackware is available on github: [https://github.com/brobr/SLACKBUILDtrees/tree/master/leo](url)

**Comment by maphew** [comment]
Thanks Rob, those would be great for my Desktop Integration sub-project as the Linux integration part is weak. See `./leo/scripts/desktop-integration.leo`. If you would please open a new issue to add what you've posted. Pull requests and patches welcome too if you're so inclined.

**Comment by brobr** [comment]
I've created a new issue from above post, hth.

**Comment by edreamleo** [comment]
On Sun, Apr 19, 2020 at 1:09 AM matt wilkie <notifications@github.com>
wrote:

> Figured it out, need to catch outer Exception and not just ImportError
> when the import is nested. Implemented in 00848c4
> <https://github.com/leo-editor/leo-editor/commit/00848c4ff2dea0a3615767fac61aaf6741edc76f>.
> I put the guard and handler in runLeo.py instead of launchLeo.py to
> support python -m leo.core.runLeo also.
>
> This is what happens now when something like pyqt isn't available:
>
> $ python -m leo.core.runLeo
> leoQt.py: can not fully import PyQt5.
> Traceback (most recent call last):
>
>   File "D:\Matt\code\leo-editor\leo\core\leoQt.py", line 52, in <module>
>     from PyQt5 import QtCore
>
> ImportError: cannot import name 'QtCore'
>
>
> name 'QtCore' is not defined
>
> *** Leo could not be started ***
> Please verify you've installed the required dependencies:https://leoeditor.com/installing.html
>
> Does this look good to you @edreamleo <https://github.com/edreamleo>?
>

Yes, so far as I understand it. Feel free to push to devel and close the
issue.

Edward
